Camera battery status publishing

hamishwillee opened the pull request:
Follows on from https://github.com/mavlink/qgroundcontrol/pull/7964

@dogmaphobic @olliw42 I don't think we should say more than this in the MAVLink spec. The QGC developer guide might additionally say what the camera widget does with such information.

Reasonable?

olliw42 commented:
on which occasion you may notice the strangeness in the naming of MAV_BATTERY_TYPE_PAYLOAD:

There is a  MAV_BATTERY_TYPE enum, but funnily MAV_BATTERY_TYPE_PAYLOAD is not member of this enum but member of the MAV_BATTERY_FUNCTION enum ... what the heck

it probably is already hardwired, so what I would do is to add a new entry MAV_BATTERY_FUNCTION_PAYLOAD and declare the MAV_BATTERY_TYPE_PAYLOAD entry as deprecated

it's not clear to me why one wouldn't add an entry MAV_BATTERY_FUNCTION_CAMERA. The whole field is IMHO actually totally obsolete for non-autopilot components, I mean the nature of the component is already revealed by  it's heartbeat, so what should this field actually tell? Wouldn't by logic extension MAV_BATTERY_TYPE_PAYLOAD mean that the camera has a payload, and that this is the status not of the camera's battery but of the camera's payload battery? MAV_BATTERY_TYPE_PAYLOAD would make sense only if the STATUS message would be send with the autopilot's ids. Here the function is in fact being the primary power source for the camera, from which perspective the function actually should be MAV_BATTERY_FUNCTION_ALL since this battery powers all of the component. The sense you should get: IMHO that's all somewhat bogus. :)

as regards your question, I would add a comment that the most relevant field is battery_remaining, and maybe time_remaining. The other fields are "interesting" but not useful for the purpose.

I'd like to mention again that IMHO it's maybe the simplest but not the best approach from a conceptual view to use the BATTERY_STATUS message and to not introduce  a CAMERA_STATUS message. I did what I did out of desperation. This PR would write it in stone.

hamishwillee commented:
Yes the naming of [MAV_BATTERY_TYPE_PAYLOAD](https://mavlink.io/en/messages/common.html#MAV_BATTERY_TYPE_PAYLOAD) is a screwup. Yes it would be good to deprecate this and add `MAV_BATTERY_FUNCTION_PAYLOAD`. That would have to have the same enum value. I believe it would work, but we'd need to generate the libraries and make sure nothing got broken. I don't have time for that right now, but if you want to create an issue would get around to it at some point. The type is a bit bogus for this case, but again, not putting brainwork into that now.

> I'd like to mention again that IMHO it's maybe the simplest but not the best approach from a conceptual view to use the BATTERY_STATUS message and to not introduce a CAMERA_STATUS message. I did what I did out of desperation. This PR would write it in stone.

I prefer using this approach to a separate CAMERA_STATUS message.

hamishwillee commented:
Re

> I would add a comment that the most relevant field is battery_remaining, and maybe time_remaining. The other fields are "interesting" but not useful for the purpose.

You're right these are the only relevant fields. Is that not obvious?

I could add/change to:

> `BATTERY_STATUS.battery_function` should be set to [MAV_BATTERY_TYPE_PAYLOAD](../messages/common.md#MAV_BATTERY_TYPE_PAYLOAD).
Generally only the `battery_remaining` field  (and possibly `time_remaining`) can be set for camera batteries.

olliw42 commented:
> That would have to have the same enum value.

I was thinking it would be a new value 5, i.e. I just would have added a new entry, and declared the other deprecated. Nothing fancy.

IMHO MAV_BATTERY_FUNCTION_ALL  makes more sense anyhow. The camera is a payload only from the autopilot's perspective, but certainly not from it's perspective.

and MAV_BATTERY_FUNCTION is generally wired anyhow

> You're right these are the only relevant fields. Is that not obvious?

well, in my experience "obvious" often happens to be quite subjective ...

I generally think that data is there to possibly make decision based on it, so I would consider it useful to tell which piece of data does what. My comment was supposed to mean that I would indicate that a GCS e.g. would primarily look for the battery_remaining and/or time_remaining. This is a different accentuation than what your sentence says. I couldn't come up with a good sentence on the quick.

hamishwillee commented:
While the camera doesn't think the battery is a payload, from the recipient point of view (GCS) it is. Docs for MAV_BATTERY_FUNCTION_ALL are "Battery supports all flight systems", which is also a big ugly. On balance I'd feel happier with either payload or stating explicitly that the types will be ignored. 

I really do think this is obvious. Below are some words. If they are not good for you, propose to leave as is.

> `BATTERY_STATUS.battery_function` should be set to `MAV_BATTERY_TYPE_PAYLOAD`.
  A GCS will typically only use the camera `battery_remaining` field (or possibly `time_remaining`); generally other fields can be set as "not supported:.

olliw42 commented:
yes, but the GCS will know it's a payload anyhow from the fact that it sees a camera heartbeat. So, totally redundant.
Note also that it's conceptually just wrong to have a component send info which is true from another components view. Just a simple example to show the general logical conflict: If you today fly your copter with one main battery without a gopro then MAV_BATTERY_FUNCTION_ALL is correct, but if you put it on tomorrow then it's not correct anymore since it's not true that all system is powered from the main battery. So, technically you have to reconfigure the autopilot each time you put on or off the gopro ... HAHAHA. You may consider this nit picking, but it's the logical flaw in here which counts not how good or bad the example is.
The MAV_BATTERY_FUNCTION is quite screwed and should - IMHO - first be reviewed and made into something reasonable, I wouldn't think it's a good idea but rather hurting later to now set things in stone.

I actually wouldn't have added ANY comment on this field in the spec !!!! If at all I would have said it's irrelevant for the purpose so can be set to best guess.

olliw42 commented:
> I really do think this is obvious.

I can't help but have to answer in a more broader sense.

I think you're not considering consequences enough, and the general behavior pattern.

Let's consider this example: The gopro is kind since it gives a percentage value, but not all cameras do. It's not too hard to imagine that a camera might be wanted to be used which only gives the voltage level. So, what to do now? The poor guy has two options, either to just set the voltage and not the percentage, and hope or assume that any recipient will do the best guess, or he works out a way to determine the percentage himself. The first however means that the responsibility to make anything useful of the data is deferred to a third party, with the consequence that this third party must adopt for very specific cases in order to just do simplest things with the data, such as issuing a warning. And all this would have to be replicated in all third parties. And it is exactly this path which leads to nightmare threads like the battery status thread (but sufficiently many other examples exist!). It very obviously would be much better from an overall perspective for the guy to work out  the percentage from the voltage. But he most likely would NOT do that since it's simpler to just set the voltage and there is no indication of which data has which relevance. But he would work that out when knowing that this is the data a third party is going to look at. 

So this little sentence is in fact most important, and anything from obvious in the sense that it will be what this guy would actually implement.

This little case is in fact just one small example of a recurring topic which demonstrated itself repeatedly, but is really somewhat surprising:

You guys need to do a better job in drafting your standard. 

One crucial aspect would be better understand how people react when they try to use MAVLink for new situations for which the standard doesn't directly apply. Most will use the messages and fields according to what they think is best, and only very few will first try to work with the MAVLink specification. But that way, you "suddenly" have various practices set in stone, which often enough are different in the different ecosystems, so that after a while it's difficult to rectify and put the train back on one trail. The proof in the pudding: Just look at the sort of corrections done in the last months. 

So, no, this little sentence is most important, and dismissing it as obvious is reflection of a broader issue. You should not let a hobbyist do what is the job of the professionals.

The sentence on the function type is in contrast quite superfluous, since it's quite irrelevant to which value this field is set. 

> Below are some words. 

The words are good. Maybe

"Other components like a GCS will typically only use the camera battery_remaining field (or possibly time_remaining); generally other fields can be set as "not supported."

hamishwillee commented:
So for a drone we let the battery provide what information it has and the GCS has a whole bunch of ways of turning this into a capacity. I think what you're saying is that while a similar approach could be used for a camera, it is actually easier for all concerned if the camera just outputs the capacity. That way even if the camera developer has do do some work to generate the capacity they know if will work for all recipients. Is that correct?

You're quite right that we need better people drafting the standard. Doing my best, but sometimes you're going to have to spell it out in tiny words.

hamishwillee commented:
OK, so I added your words for the GCS. Makes sense/certainly easier for all concerned.

It does not really matter what you put for `BATTERY_STATUS.battery_function` because the GCS isn't going to use that. I have removed that line. I agree `MAV_BATTERY_FUNCTION_ALL` is "most correct"
`MAV_BATTERY_FUNCTION_ALL` but not as it is documented so will follow up with a separate PR. 

@dogmaphobic I'm merging this. Let me know if you think there should be further edits.

olliw42 commented:
THX, sir :)
